File: examples/latest/navigate/android/TruckGuidance/app/src/main/java/com/here/truckguidance/SpeedView.java
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.truckguidance;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.util.TypedValue;
import android.view.View;

// A simple view to show the current speed limit or current driving speed.
public class SpeedView extends View {

    private Paint paint;
    private String speedText = "";
    private String labelText = "";

    private int yMarginInDP = 5;
    private int textSizeInDP = 15;
    public int circleColor = Color.RED;

    // The dimensions of the rectangle that holds all content.
    public int xInDP = 0;
    public int yInDP = 0;
    private int wInDP = 50;
    private int hInDP = wInDP + yMarginInDP + textSizeInDP;

    public SpeedView(Context context) {
        super(context);
        paint = new Paint();
    }

    public int getWidthInDP() {
        return wInDP;
    }

    public int getHeightInDP() {
        return hInDP;
    }

    public void redraw() {
        // Request a redraw.
        invalidate();
    }

    public void setSpeedLimit(String text) {
        this.speedText = text;
        redraw();
    }

    public void setLabel(String text) {
        this.labelText = text;
        redraw();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // Draw the outer circle.
        paint.setColor(circleColor);
        canvas.drawCircle(getCircleCenterX(), getCircleCenterY(), getRadius(), paint);

        // Draw the inner circle.
        paint.setColor(Color.WHITE);
        float innerCircleRadius = getPixels((int) (wInDP * 0.37));
        canvas.drawCircle(getCircleCenterX(), getCircleCenterY(), innerCircleRadius, paint);

        drawSpeedValue(canvas);
        drawLabel(canvas);
    }

    private float getRadius() {
        return getPixels(wInDP / 2);
    }

    private float getCircleCenterX() {
        return getRadius() + getPixels(xInDP);
    }

    private float getCircleCenterY() {
        return getRadius() + getPixels(yInDP) + getPixels(textSizeInDP);
    }

    private void drawSpeedValue(Canvas canvas) {
        if (speedText.isEmpty()) {
            // Nothing to draw.
            return;
        }

        // Draw the speed text centered in circle.
        paint.setColor(Color.BLACK);
        paint.setTextSize(getPixels(textSizeInDP));
        paint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));
        float textWidth = paint.measureText(speedText);
        float textCenterX = getCircleCenterX() - (textWidth / 2);
        float textCenterY = getCircleCenterY() - (paint.descent() + paint.ascent()) / 2;
        canvas.drawText(speedText, textCenterX, textCenterY, paint);
    }

    private void drawLabel(Canvas canvas) {
        if (labelText.isEmpty()) {
            // Nothing to draw.
            return;
        }

        // Draw the label centered on top of the circle.
        paint.setColor(Color.BLACK);
        paint.setTextSize(getPixels(textSizeInDP));
        paint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));
        float textWidth = paint.measureText(labelText);
        float radius = getPixels(wInDP / 2);
        float circleCenterY = getCircleCenterY() - radius - getPixels(yMarginInDP * 2);
        float textCenterX = getCircleCenterX() - (textWidth / 2);
        float textCenterY = circleCenterY - (paint.descent() + paint.ascent()) / 2;
        canvas.drawText(labelText, textCenterX, textCenterY, paint);
    }

    // Get pixel size for this particular device based on its display metrics.
    // See https://developer.android.com/training/multiscreen/screendensities
    private int getPixels(int dp) {
        int pixels = Math.round(TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, dp, getResources().getDisplayMetrics()));
        return pixels;
    }
}

File: examples/latest/navigate/flutter/truck_guidance_app/lib/speed_view.dart
/*
 * Copyright (C) 2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
import 'package:flutter/material.dart';

/// A custom widget that mimics the SpeedView from Java.
/// It draws a circle with an inner circle and overlays centered text.
class SpeedView extends StatelessWidget {
  final String label;
  final String speed;
  final Color circleColor;

  const SpeedView({Key? key, required this.label, required this.speed, required this.circleColor}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    // The Java code uses wInDP=50, yMarginInDP=5, textSizeInDP=15.
    // Here we set the size accordingly.
    return CustomPaint(
      size: const Size(50, 50 + 5 + 15),
      painter: SpeedViewPainter(label: label, speed: speed, circleColor: circleColor),
    );
  }
}

class SpeedViewPainter extends CustomPainter {
  final String label;
  final String speed;
  final Color circleColor;

  SpeedViewPainter({required this.label, required this.speed, required this.circleColor});

  @override
  void paint(Canvas canvas, Size size) {
    // Dimensions (in logical pixels) based on the Java values.
    double wInDP = 50;
    double yMarginInDP = 5;
    double textSizeInDP = 15;
    double radius = wInDP / 2;
    // Circle center is shifted downward by textSizeInDP.
    Offset center = Offset(radius, radius + textSizeInDP);

    Paint paint = Paint()..isAntiAlias = true;

    // Draw outer circle.
    paint.color = circleColor;
    canvas.drawCircle(center, radius, paint);

    // Draw inner circle.
    paint.color = Colors.white;
    double innerCircleRadius = wInDP * 0.37;
    canvas.drawCircle(center, innerCircleRadius, paint);

    // Draw speed text centered in circle.
    if (speed.isNotEmpty) {
      TextSpan speedSpan = TextSpan(
        style: TextStyle(color: Colors.black, fontSize: textSizeInDP, fontWeight: FontWeight.bold),
        text: speed,
      );
      TextPainter speedTP = TextPainter(text: speedSpan, textAlign: TextAlign.center, textDirection: TextDirection.ltr);
      speedTP.layout();
      Offset speedOffset = Offset(center.dx - speedTP.width / 2, center.dy - speedTP.height / 2);
      speedTP.paint(canvas, speedOffset);
    }

    // Draw label centered above the circle.
    if (label.isNotEmpty) {
      TextSpan labelSpan = TextSpan(
        style: TextStyle(color: Colors.black, fontSize: textSizeInDP, fontWeight: FontWeight.bold),
        text: label,
      );
      TextPainter labelTP = TextPainter(text: labelSpan, textAlign: TextAlign.center, textDirection: TextDirection.ltr);
      labelTP.layout();
      double labelY = center.dy - radius - yMarginInDP * 2 - labelTP.height / 2;
      Offset labelOffset = Offset(center.dx - labelTP.width / 2, labelY);
      labelTP.paint(canvas, labelOffset);
    }
  }

  @override
  bool shouldRepaint(covariant SpeedViewPainter oldDelegate) {
    return oldDelegate.label != label || oldDelegate.speed != speed || oldDelegate.circleColor != circleColor;
  }
}
